core/model/__init__.py

```diff
@@ -334,38 +334,40 @@ def initialize(cls, url, initialize_data=True, initialize_schema=True):
             return engine, engine.connect()
 
         engine = cls.engine(url)
-        if initialize_schema:
-            cls.initialize_schema(engine)
-        connection = engine.connect()
-
-        # Check if the recursive equivalents function exists already.
-        query = (
-            select([literal_column("proname")])
-            .select_from(table("pg_proc"))
-            .where(literal_column("proname") == "fn_recursive_equivalents")
-        )
-        result = connection.execute(query)
-        result = list(result)
-
-        # If it doesn't, create it.
-        if not result and initialize_data:
-            resource_file = os.path.join(
-                cls.resource_directory(), cls.RECURSIVE_EQUIVALENTS_FUNCTION
+        with engine.connect() as connection:
+            tx = connection.begin()
```

Comment:
This was an especially tricky bit.  We need a way to lock the database on startup when no schema exists.  Since there are no cm tables yet,  I'm using postges internal stored procedure metadata table. We need to open a transaction and close it only when we're done with our initialization.  The reason is that without a transaction,  I believe line 342 (`cls.initialize_schema(connection)`) will spawn a separate transaction causing a deadlock.  I'm not totally clear on the mechanics here - the sqlalchemy code is pretty dense.

Reply:
Most of the changes below are just side effects of the indent required by the transaction and with statement.